src/incremental.jl

```diff
@@ -0,0 +1,183 @@
+
+Extracts the value for `flag` from the current julia cmd
+"""
+function extract_flag(flag, jl_cmd = Base.julia_cmd())
+    for elem in jl_cmd.exec
+        if startswith(elem, flag)
+            return elem
+        end
+    end
+    return nothing
+end
+
+function extract_flag_value(flag, jl_cmd = Base.julia_cmd())
+    val = extract_flag(flag, jl_cmd)
+    replace(val, flag => "")
+end
+current_systemimage() = extract_flag_value("-J")
+
+"""
+    add_command!(cmd, ignore, jl_cmd, flag, value)
+
+Adds a command to `cmd` - either choosing the julia default for value == nothing,
+or adds value!. If ignore = true, and command not in jl_cmd && value == nothing,
+command gets not added.
+"""
+function add_command!(cmd, ignore, jl_cmd, flag, value)
+    new_cmd = if value == nothing
+        extract_flag(flag, jl_cmd)
+    else
+        short = !startswith(flag, "--")
+        string(flag, short ? "" : "=", value)
+    end
+    if new_cmd == nothing
+        ignore || @warn("Flag $flag not present in julia-cmd, but is set to nothing - can't add the flag")
+    else
+        push!(cmd, new_cmd)
+    end
+end
+
+"""
+    run_julia(
+        code::String;
+        g = nothing, O = nothing, output_o = nothing, J = nothing,
+        startup_file = "no"
+    )
+
+Runs `code` in a new julia command!
+You can overwrite any julia command line flag by setting it to a value.
+If nothing is chosen, it will default to the value of the current julia process.
+"""
+function run_julia(code::String; kw...)
+    run(julia_code_cmd(code; kw...))
+end
+
+function julia_code_cmd(
+        code::String;
+        g = nothing, O = nothing, output_o = nothing, J = nothing,
+        startup_file = nothing, trace_compile = nothing, compile = nothing
+    )
+
+    jl_cmd = Base.julia_cmd()
+    cmd = `$(jl_cmd.exec[1])` # extract julia exe
+
+    add_command!(cmd.exec, false, jl_cmd, "-O", O)
+    add_command!(cmd.exec, false, jl_cmd, "-J", J)
+
+    add_command!(cmd.exec, true, jl_cmd, "-g", g)
+    add_command!(cmd.exec, true, jl_cmd, "--output-o", output_o)
+    add_command!(cmd.exec, true, jl_cmd, "--startup-file", startup_file)
+    add_command!(cmd.exec, true, jl_cmd, "--compile", compile)
+    add_command!(cmd.exec, true, jl_cmd, "--trace-compile", trace_compile)
+    # for better debug, let's not make a tmp file and immediately delete it!
+    file = sysimg_folder("run_julia_code.jl")
+    open(io-> println(io, code), file, "w")
+    push!(cmd.exec, file)
+    cmd
+end
+
+
+"""
+    compile_incremental(
+        toml_path::String, snoopfile::String;
+        force = false, reuse = false, verbose = true,
+        debug = false, cc_flags = nothing
+    )
+
+    Extract all calls from `snoopfile` and ahead of time compiles them
+    incrementally into the current system image.
+    `force = true` will replace the old system image with the new one.
+    The `toml` will need to contain all packages that `snoopfile` uses explicitely.
```

Comment:
```suggestion
    The argument `toml_path` should contain a project file of the packages that `snoopfile` explicitly uses.
```

Should there be a manifest too?

Reply:
Good question. Isn't that used implicitly by activate when stored in the same folder?